instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/field/VirtualField.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.instrumentation.api.field;
+
+import io.opentelemetry.instrumentation.api.internal.RuntimeVirtualFieldSupplier;
+import java.util.function.Supplier;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+/**
+ * Represents a "virtual" field of type {@code F} that is added to type {@code T} in the runtime.
+ *
+ * <p>Field values are weakly referenced and will be garbage collected when their owner instance is
+ * collected.
+ *
+ * @param <T> The type that will contain the new virtual field.
+ * @param <F> The field type that'll be added to {@code T}.
+ */
+// we're using an abstract class here so that we can call static find() in pre-jdk8 advice classes
+public abstract class VirtualField<T, F> {
+
+  /**
+   * Finds a {@link VirtualField} instance for given {@code type} and {@code fieldType}.
+   *
+   * <p>Conceptually this can be thought of as a map lookup to fetch a second level map given {@code
+   * type}.
+   *
+   * <p>In runtime, when using the javaagent, the <em>calls</em> to this method are rewritten to
+   * something more performant while injecting advice into a method.
+   *
+   * <p>When using this method outside of Advice method, the {@link VirtualField} should be looked
+   * up once and stored in a field to avoid repeatedly calling this method.
+   *
+   * @param type The type that will contain the new virtual field.
+   * @param fieldType The field type that'll be added to {@code type}.
+   */
+  public static <U extends T, T, F> VirtualField<U, F> find(Class<T> type, Class<F> fieldType) {
+    return RuntimeVirtualFieldSupplier.get().find(type, fieldType);
+  }
+
+  /** Gets the value of this virtual field. */
+  @Nullable
+  public abstract F get(T object);
+
+  /** Sets the new value of this virtual field. */
+  public abstract void set(T object, @Nullable F fieldValue);
+
+  /**
+   * Sets the new value of this virtual field if the current value is {@code null} or absent.
```

Comment:
I would remove the or absent part as we really only check whether the field is `null`

Reply:
Good point.
Now I'm starting to wonder whether the method should be named `setIfNullAndGet`...

Reply:
I like `setIfNull`

Reply:
👍

Reply:
i was thinking `setIfNull` without the `AndGet`, similar to `Map.putIfAbsent`, but now i realize the potential confusion because `putIfAbsent` returns the previous value or `null` if there was none, while this method always returns the stored value.

maybe:

`void setIfNull(T object, F fieldValue)`

and

`F computeIfNull(T object, Supplier<F> fieldValueSupplier)`

as it looks like we don't (currently at least) use the return value from the first method, though now I'm wondering if that method is even needed, i.e. will anything break if those callers just use `set` (and if the "set if null" behavior really does matter, they could call `get` first, as there's nothing atomic anyways if they aren't using the return value)

Reply:
> though now I'm wondering if that method is even needed, i.e. will anything break if those callers just use `set` (and if the "set if null" behavior really does matter, they could call `get` first, as there's nothing atomic anyways if they aren't using the return value)

Most of the `setIfNull` usages look like this method is used instead of plain `set` "just in case something happens"; even though there seems to be no real chance of the field being non-null. I can take a look at them and maybe remove the method in a follow-up PR.